homeassistant/components/mazda/__init__.py

```diff
@@ -59,6 +65,77 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
         _LOGGER.error("Error occurred during Mazda login request: %s", ex)
         raise ConfigEntryNotReady from ex
 
+    async def async_handle_service_call(service_call=None):
+        """Handle a service call."""
+        # Get device entry from device registry
+        dev_reg = device_registry.async_get(hass)
+        device_id = service_call.data.get("device_id")
+        device_entry = dev_reg.async_get(device_id)
+
+        # Get vehicle VIN from device identifiers
+        mazda_identifiers = [
+            identifier
+            for identifier in device_entry.identifiers
+            if identifier[0] == DOMAIN
+        ]
+        vin_identifier = next(iter(mazda_identifiers))
+        vin = vin_identifier[1]
+
+        # Get vehicle ID and API client from hass.data
+        vehicle_id = 0
+        api_client = None
+        for entry_data in hass.data[DOMAIN].values():
+            for vehicle in entry_data[DATA_VEHICLES]:
+                if vehicle["vin"] == vin:
+                    vehicle_id = vehicle["id"]
+                    api_client = entry_data[DATA_CLIENT]
+
+        if vehicle_id == 0 or api_client is None:
+            raise HomeAssistantError("Vehicle ID not found")
+
+        api_method = getattr(api_client, service_call.service)
+        try:
+            if service_call.service == "send_poi":
+                latitude = service_call.data.get("latitude")
+                longitude = service_call.data.get("longitude")
+                poi_name = service_call.data.get("poi_name")
+                await api_method(vehicle_id, latitude, longitude, poi_name)
+            else:
+                await api_method(vehicle_id)
+        except Exception as ex:
+            _LOGGER.exception("Error occurred during Mazda service call: %s", ex)
+            raise HomeAssistantError(ex) from ex
```

Comment:
If we're raising an exception we shouldn't also log the stack trace. Include the log message in the exception argument instead.